File: src/ruby/prime_and_composite_tautology.rb
require ".\\core.rb"

if ARGV.length!= 1 then
    print "usage: ruby prime_and_composite_tautology.rb number."
    exit 1
end

bit_width=ARGV[0]

if !bit_width.match(/^(\d)+$/) then
    print "usage: ruby prime_and_composite_tautology.rb number."
    exit 1
end

bit_width=bit_width.to_i

conditions=[
    IsPrime.new("target",bit_width),
    IsComposite.new("target",bit_width),
    Input_Equals_Number.new("One_NBit_#{Z(bit_width)}",1,bit_width),
    Input_Equals_Number.new("One_NBit_#{Z(bit_width*2)}",1,bit_width*2),
    " -<Zero_1Bit_#{Z(1)}> 0 "
]

generate_cnf(conditions,"prime_and_composite_tautology_#{bit_width}.cnf")

File: src/cpp/prime_and_composite_tautology.cpp
#include "core.hpp"
#include <iostream>
#include <regex>
#include <vector>
#include <string>
int main(int argc, char* argv[]) {
    if (argc != 2) {
        std::cout << "usage: prime_and_composite_tautology number." << std::endl;
        return 1;
    }
    std::string bit_width_str = argv[1];
    if (!std::regex_match(bit_width_str, std::regex("^\\d+$"))) {
        std::cout << "usage: prime_and_composite_tautology number." << std::endl;
        return 1;
    }
    int bit_width = std::stoi(bit_width_str);
    std::vector<std::string> conditions;
    {
        IsPrime is_prime("target", bit_width, bit_width);
        auto v = is_prime.expand();
        conditions.insert(conditions.end(), v.begin(), v.end());
    }
    {
        IsComposite is_composite("target", bit_width);
        auto v = is_composite.expand();
        conditions.insert(conditions.end(), v.begin(), v.end());
    }
    {
        Input_Equals_Number ien1("One_NBit_" + Z(bit_width), 1, bit_width);
        auto v = ien1.expand();
        conditions.insert(conditions.end(), v.begin(), v.end());
    }
    {
        Input_Equals_Number ien2("One_NBit_" + Z(bit_width*2), 1, bit_width*2);
        auto v = ien2.expand();
        conditions.insert(conditions.end(), v.begin(), v.end());
    }
    conditions.push_back(" -<Zero_1Bit_" + Z(1) + "> 0 ");
    generate_cnf(conditions, "prime_and_composite_tautology_" + std::to_string(bit_width) + ".cnf");
    return 0;
} 